### PR TITLE
[SPARK-8579] [SQL] support arbitrary object in UnsafeRow

### PR DESCRIPTION
**Pull request by davies**
This PR brings arbitrary object support in UnsafeRow (both in grouping key and aggregation buffer).

Two object pools will be created to hold those non-primitive objects, and put the index of them into UnsafeRow. In order to compare the grouping key as bytes, the objects in key will be stored in a unique object pool, to make sure same objects will have same index (used as hashCode).

For StringType and BinaryType, we still put them as var-length in UnsafeRow when initializing for better performance. But for update, they will be an object inside object pools (there will be some garbages left in the buffer).

BTW: Will create a JIRA once issue.apache.org is available.

cc @JoshRosen @rxin

### CODE REVIEW

**Comment by davies** [comment]
@JoshRosen This one is ready to review, thanks!

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java [hunk 1]
```diff
@@ -70,67 +92,38 @@
...
       TaskMemoryManager memoryManager,
       int initialCapacity,
       boolean enablePerfMetrics) {
-    this.emptyAggregationBuffer =
-      convertToUnsafeRow(emptyAggregationBuffer, aggregationBufferSchema);
-    this.aggregationBufferSchema = aggregationBufferSchema;
-    this.groupingKeyToUnsafeRowConverter = new UnsafeRowConverter(groupingKeySchema);
-    this.groupingKeySchema = groupingKeySchema;
-    this.map = new BytesToBytesMap(memoryManager, initialCapacity, enablePerfMetrics);
+    this.initProjection = initProjection;
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: wouldn't this decrease performance?

Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: This is necessary, for example, you need to create a `OpenHashSet` for every grouping key. Also, this is only called when having object in the initial value.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 2]
```diff
@@ -171,8 +113,56 @@ private void setNotNullAt(int i) {
   }
 
   @Override
-  public void update(int ordinal, Object value) {
-    throw new UnsupportedOperationException();
+  public void update(int i, Object value) {
+    if (value == null) {
```
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: can you add some comments inline on what is happening? It is pretty hard to follow for this function.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ObjectPool.scala [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+package org.apache.spark.sql.catalyst.expressions
+
+/**
+ * A object pool stores a collection of objects in array, then they can be referenced by the
+ * pool plus an index.
+ *
+ * @param capacity the initial capacity
+ */
+class ObjectPool(var capacity: Int) {
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ObjectPool.scala]: rewrite this in java? since it is dependent of a java file

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/UnsafeRowConverterSuite.scala [hunk 4]
```diff
@@ -66,14 +72,59 @@ class UnsafeRowConverterSuite extends SparkFunSuite with Matchers {
...
     numBytesWritten should be (sizeRequired)
 
     val unsafeRow = new UnsafeRow()
-    unsafeRow.pointTo(buffer, PlatformDependent.LONG_ARRAY_OFFSET, fieldTypes.length, null)
+    val pool = new ObjectPool(10)
+    unsafeRow.pointTo(buffer, PlatformDependent.LONG_ARRAY_OFFSET, fieldTypes.length, pool)
     unsafeRow.getLong(0) should be (0)
     unsafeRow.getString(1) should be ("Hello")
-    unsafeRow.getBinary(2) should be ("World".getBytes)
+    unsafeRow.get(2) should be ("World".getBytes)
```
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/UnsafeRowConverterSuite.scala]: can we stop using should matchers for simple equality comparisons? just use ===

should matchers are pretty complicated and more verbose to read

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ObjectPool.scala (hunk 3 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ObjectPool.scala]: I'd like use Scala as possible, what's the benefit if it's written in Java?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 2 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: done

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java [hunk 5]
```diff
@@ -38,16 +38,28 @@
    * An empty aggregation buffer, encoded in UnsafeRow format. When inserting a new key into the
    * map, we copy this buffer and use it as the value.
    */
-  private final byte[] emptyAggregationBuffer;
+  private final byte[] emptyBuffer;
 
-  private final StructType aggregationBufferSchema;
+  /**
+   * A empty row used by `initProjection`
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: A -> An

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java [hunk 6]
```diff
@@ -55,9 +67,19 @@
   private final BytesToBytesMap map;
 
   /**
+   * A object pool for these in keys.
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: How about "An object pool for objects that are used in grouping keys"

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java [hunk 7]
```diff
@@ -55,9 +67,19 @@
...
+   */
+  private final UniqueObjectPool keyPool;
+
+  /**
+   * A object pool for those in aggregation buffers.
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: How about "An object pool for objects that are used in aggregation buffers"

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java [hunk 8]
```diff
@@ -55,9 +67,19 @@
...
+   */
+  private final ObjectPool bufferPool;
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: I think `bufferPool` is a slightly confusing name, since I think that readers may confuse it with a memory page buffer pool used in databases.  Instead, maybe naming it something like "valueObjectPool" or "aggregationBufferObjectPool" would make this clearer.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java [hunk 9]
```diff
@@ -55,9 +67,19 @@
   private final BytesToBytesMap map;
 
   /**
+   * A object pool for these in keys.
+   */
+  private final UniqueObjectPool keyPool;
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: Can we bet a bit more verbose here and call this `keyObjectPool`?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java (hunk 8 above)
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: `bufferObjectPool` would be fine, too.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java [hunk 10]
```diff
@@ -70,67 +92,38 @@
...
+    this.keyPool = new UniqueObjectPool(100);
+    this.bufferPool = new ObjectPool(initialCapacity);
+
+    InternalRow initRow = initProjection.apply(emptyRow);
+    this.emptyBuffer = new byte[bufferConverter.getSizeRequirement(initRow)];
+    int writtenLength = bufferConverter.writeRow(
+      initRow, emptyBuffer, PlatformDependent.BYTE_ARRAY_OFFSET, bufferPool);
+    assert (writtenLength == emptyBuffer.length): "Size requirement calculation was wrong!";
+    // re-use the empty buffer only when there is no object saved in pool.
+    reuseEmptyBuffer = bufferPool.size() == 0;
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: `reuseEmptyBuffer` is only assigned to once, which seems like a bug.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 11]
```diff
@@ -48,7 +38,15 @@
  * primitive types, such as long, double, or int, we store the value directly in the word. For
  * fields with non-primitive or variable-length values, we store a relative offset (w.r.t. the
  * base address of the row) that points to the beginning of the variable-length field, and length
- * (they are combined into a long).
+ * (they are combined into a long). For other objects, they are stored in a pool, the indexes of
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: The tricky part of this patch seems to be handling fast hashing / equality checks for UnsafeRows that contain objects.  Can we add a paragraph here to explain how that works / why that's safe?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 12]
```diff
@@ -48,7 +38,15 @@
...
+ * them are hold in the the word.
+ *
+ * For non-primitive types, the word of a field could be:
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Thanks for adding this comment; it's very helpful.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 13]
```diff
@@ -171,8 +122,64 @@ private void setNotNullAt(int i) {
   }
 
   @Override
-  public void update(int ordinal, Object value) {
-    throw new UnsupportedOperationException();
+  public void update(int i, Object value) {
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: There seems to be an unstated assumption that this method will only be called to update String, Binary, or Object-valued columns and not primitive columns.  Can you add a comment stating this explicitly?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 14]
```diff
@@ -171,8 +122,64 @@ private void setNotNullAt(int i) {
...
+    if (value == null) {
+      if (!isNullAt(i)) {
+        // remove the old value from pool
+        long idx = getLong(i);
+        if (idx <= 0) {
+          // this is the index of old value in pool, remove it
+          pool.replace((int)-idx, null);
+        } else {
+          // there will be some garbage left (UTF8String or byte[])
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: I guess one consequence of this is that equality comparisons of modified UnsafeRows won't work as expected if we just compare the bytes, since creating an UnsafeRow with a null value for a particular string / byte column produces a row with different bytes than if we had created a row with a non-null value for the column then subsequently set it to null.

This wasn't a problem before because we didn't allow updates to string-valued columns, but it might be a problem here for hashing / comparison of rows.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 15]
```diff
@@ -171,8 +122,64 @@ private void setNotNullAt(int i) {
...
+    } else {
+      // there is an old value, check the type, then replace it or update it
+      long v = getLong(i);
+      if (v <= 0) {
+        // it's the index in the pool, replace old value with new one
+        int idx = (int)-v;
+        pool.replace(idx, value);
+      } else {
+        // old value is UTF8String or byte[], try to reuse the space
+        boolean is_string;
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: This should probably be camel-cased.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 16]
```diff
@@ -171,8 +122,64 @@ private void setNotNullAt(int i) {
...
+        if (newBytes.length <= oldLength) {
+          // the new value can fit in the old buffer, re-use it
+          PlatformDependent.copyMemory(
+            newBytes,
+            PlatformDependent.BYTE_ARRAY_OFFSET,
+            baseObject,
+            baseOffset + offset,
+            newBytes.length);
+          long flag = is_string ? 1L << (OFFSET_BITS * 2) : 0L;
+          setLong(i, flag | (((long)offset) << OFFSET_BITS) | (long)newBytes.length);
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: AFAIK we don't have a style guide for Java code, but I think we should put spaces after the casts.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 17]
```diff
@@ -230,27 +237,32 @@ public int size() {
...
+      int idx = (int)-v;
+      return pool.get(idx);
     } else {
-      throw new UnsupportedOperationException();
+      boolean is_string = (v >> (OFFSET_BITS * 2)) > 0;
+      int offset = (int)((v >> OFFSET_BITS) & Integer.MAX_VALUE);
+      int size = (int)(v & Integer.MAX_VALUE);
+      final byte[] bytes = new byte[size];
+      PlatformDependent.copyMemory(
+              baseObject,
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: I think these lines are over-indented.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 18]
```diff
@@ -230,27 +237,32 @@ public int size() {
   }
 
   @Override
-  public StructType schema() {
-    return schema;
-  }
-
-  @Override
   public Object get(int i) {
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Similar to my comments on the other methods, do you mind adding a comment to `get()` to explain which types we expected it to be called on?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/UnsafeRowConverter.scala [hunk 19]
```diff
@@ -257,12 +227,10 @@ private abstract class BytesUnsafeColumnWriter extends UnsafeColumnWriter {
     ByteArrayMethods.roundNumberOfBytesToNearestWord(numBytes)
   }
 
-  override def write(
-      source: InternalRow,
-      target: UnsafeRow,
-      column: Int,
-      appendCursor: Int): Int = {
-    val offset = target.getBaseOffset + appendCursor
+  def isString: Boolean
```
Comment (JoshRosen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/UnsafeRowConverter.scala]: I think this can be a `protected[this]` val instead.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 2 above)
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Can we add a unit test which hits this branch?  I didn't see it get hit by the existing tests when I ran them with IntellliJ's coverage.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 20]
```diff
@@ -171,8 +122,64 @@ private void setNotNullAt(int i) {
...
+        } else {
+          // there will be some garbage left (UTF8String or byte[])
+        }
+        setNullAt(i);
+      }
+      return;
+    }
+
+    if (isNullAt(i)) {
+      // there is not an old value, put the new value into pool
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Looks like this branch also got missed in test coverage.

**Comment by JoshRosen** [comment]
Left a number of small comments. I still want to spend a bit more time to carefully think through how equality and hashing of UnsafeRows work to see whether there are any corner-cases that we should add tests for. My review is getting pre-empted by some other stuff, though, so I'll loop back later to add more comments.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java (hunk 8 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: The type is already `ObjectPool`,  having the full name in the variable name is too verbose, I think.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java (hunk 10 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: change it to final

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 14 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: We only compare UnsafeRow as bytes in key ofBytesToBytesMap, the key cannot be changed in this case. Otherwise, we will use the general way to compare UnsafeRow with other Rows (calling InternalRow.equals() and hashCode()).

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 20 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: covered

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 11 above)
Comment (davies) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: added

**Comment by davies** [comment]
@JoshRosen Could you take a second pass? I should had addressed all the previous comments.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMapSuite.scala [hunk 21]
```diff
@@ -33,10 +34,10 @@ class UnsafeFixedWidthAggregationMapSuite
   with Matchers
   with BeforeAndAfterEach {
 
-  import UnsafeFixedWidthAggregationMap._
-
   private val groupKeySchema = StructType(StructField("product", StringType) :: Nil)
   private val aggBufferSchema = StructType(StructField("salePrice", IntegerType) :: Nil)
+  private def emptyProjection: Projection =
+    GenerateProjection.generate(Seq(Literal(0)), Seq(AttributeReference("price", IntegerType)()))
   private def emptyAggregationBuffer: InternalRow = InternalRow(0)
```
Comment (JoshRosen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMapSuite.scala]: IntelliJ says that this is now unused.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java (hunk 10 above)
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: There's a corner-case that I'm worried about where we have an object-valued column that contains `null` in the empty buffer.  I'm concerned that this won't properly trigger the object pool size increase, which may cause us to inappropriately re-use the empty buffer.  I'm playing around with it now in IntelliJ to see if I can get this to happen. If so, I'll send a PR to fix it.

Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeFixedWidthAggregationMap.java]: Looked more closely and this is fine.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 22]
```diff
@@ -44,7 +36,20 @@
  * primitive types, such as long, double, or int, we store the value directly in the word. For
  * fields with non-primitive or variable-length values, we store a relative offset (w.r.t. the
  * base address of the row) that points to the beginning of the variable-length field, and length
- * (they are combined into a long).
+ * (they are combined into a long). For other objects, they are stored in a pool, the indexes of
+ * them are hold in the the word.
+ *
+ * In order to support fast hashing and equality checks for UnsafeRows that contain objects
+ * when used as grouping key in BytesToBytesMap, we put the objects in an UniqueObjectPool to make
+ * sure all the keyhave the same index for same object, then we can hash/compare the objects by
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Typo: `keyhave`

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/UniqueObjectPool.java [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+
+/**
+ * An unique object pool stores a collection of unique objects in it.
+ */
+public class UniqueObjectPool extends ObjectPool {
+
+  /**
+   * A hash map from objects to their indexes in the array.
+   */
+  private  HashMap<Object, Integer> objIndex;
```
Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/UniqueObjectPool.java]: Minor style nit: should only be one space after `private`.

Comment (JoshRosen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/UniqueObjectPool.java]: It looks like this is an append-only map where entries are immutable, so we might be able to use core's `AppendOnlyMap` for this instead.  OTOH, AppendOnlyMap doesn't cache hashCode values and this map does.  Probably fine to leave as is.

**Comment by JoshRosen** [comment]
LGTM overall.

**Comment by davies** [comment]
@JoshRosen Thanks, I'm merging this into master.
